File: frontend/src/types/tiers.ts
/**
 * Canonical tier definitions — MUST mirror backend/app/core/tiers.py exactly.
 * Single source of truth. Per CLAUDE.md Pricing Model v4 (FINAL).
 * Do not add, rename, or reorder without Phil Bölke approval.
 *
 * Regenerate via: openapi-typescript http://localhost:8000/openapi.json
 */

export const Tier = {
  FREE: "free",
  SPARK: "spark",
  CREATOR: "creator",
  ARTIST: "artist",
  STUDIO_PRO: "studio_pro",
  ENTERPRISE: "enterprise",
} as const;

export type Tier = typeof Tier[keyof typeof Tier];

export const TIER_RANK: Record<Tier, number> = {
  [Tier.FREE]: 0,
  [Tier.SPARK]: 1,
  [Tier.CREATOR]: 2,
  [Tier.ARTIST]: 3,
  [Tier.STUDIO_PRO]: 4,
  [Tier.ENTERPRISE]: 5,
};

export const TIER_DISPLAY_NAME: Record<Tier, string> = {
  [Tier.FREE]: "Free",
  [Tier.SPARK]: "Spark",
  [Tier.CREATOR]: "Creator",
  [Tier.ARTIST]: "Artist",
  [Tier.STUDIO_PRO]: "Studio Pro",
  [Tier.ENTERPRISE]: "Enterprise",
};

export const TIER_PRICE_USD_MONTHLY: Record<Tier, number | null> = {
  [Tier.FREE]: 0,
  [Tier.SPARK]: 9,
  [Tier.CREATOR]: 29,
  [Tier.ARTIST]: 59,
  [Tier.STUDIO_PRO]: 149,
  [Tier.ENTERPRISE]: null, // custom, ~$499+
};

export const TIER_DOWNLOADS_PER_MONTH: Record<Tier, number | "unlimited"> = {
  [Tier.FREE]: 0,
  [Tier.SPARK]: 50,
  [Tier.CREATOR]: 10,
  [Tier.ARTIST]: 25,
  [Tier.STUDIO_PRO]: 75,
  [Tier.ENTERPRISE]: "unlimited",
};

/** Returns true if `tier` has at least the rank of `minimum`. */
export function tierGte(tier: Tier | string, minimum: Tier | string): boolean {
  const t = TIER_RANK[tier as Tier];
  const m = TIER_RANK[minimum as Tier];
  if (t === undefined || m === undefined) return false;
  return t >= m;
}

/** Tier feature gates (frontend convenience — backend enforces). */
export const TIER_FEATURES = {
  canDownload: (t: Tier) => tierGte(t, Tier.SPARK),
  canUseStems: (t: Tier) => tierGte(t, Tier.CREATOR),
  canUseClaude: (t: Tier) => tierGte(t, Tier.CREATOR),
  canUseDawPlugin: (t: Tier) => tierGte(t, Tier.ARTIST),
  canUseAIE: (t: Tier) => tierGte(t, Tier.ARTIST),
  canUseDistribution: (t: Tier) => tierGte(t, Tier.ARTIST),
  canUseAtmos: (t: Tier) => tierGte(t, Tier.STUDIO_PRO),
  canUseDDEX: (t: Tier) => tierGte(t, Tier.STUDIO_PRO),
  canUseDDP: (t: Tier) => tierGte(t, Tier.STUDIO_PRO),
  canUseVinyl: (t: Tier) => tierGte(t, Tier.STUDIO_PRO),
  canUseCollaboration: (t: Tier) => tierGte(t, Tier.STUDIO_PRO),
  canUseCustomRainNet: (t: Tier) => tierGte(t, Tier.ENTERPRISE),
  canUseWhiteLabel: (t: Tier) => tierGte(t, Tier.ENTERPRISE),
} as const;

File: backend/app/core/tiers.py
"""Canonical tier definitions. Single source of truth — shared with frontend via openapi-typescript.
Per CLAUDE.md Pricing Model v4 (FINAL). Do not add, rename, or reorder without Phil Bölke approval.
"""
from __future__ import annotations
from enum import Enum


class Tier(str, Enum):
    FREE = "free"
    SPARK = "spark"
    CREATOR = "creator"
    ARTIST = "artist"
    STUDIO_PRO = "studio_pro"
    ENTERPRISE = "enterprise"


# Rank ordering for tier_gte comparisons. Must match Tier enum order.
TIER_RANK: dict[Tier, int] = {
    Tier.FREE: 0,
    Tier.SPARK: 1,
    Tier.CREATOR: 2,
    Tier.ARTIST: 3,
    Tier.STUDIO_PRO: 4,
    Tier.ENTERPRISE: 5,
}


# Rate limits per CLAUDE.md tier architecture. Requests/hour.
TIER_RATE_LIMITS: dict[Tier, str] = {
    Tier.FREE: "20/hour",
    Tier.SPARK: "100/hour",
    Tier.CREATOR: "500/hour",
    Tier.ARTIST: "1500/hour",
    Tier.STUDIO_PRO: "5000/hour",
    Tier.ENTERPRISE: "50000/hour",
}


# Celery queue routing by tier (see worker.py task_routes).
TIER_QUEUE: dict[Tier, str] = {
    Tier.FREE: "cpu_standard",
    Tier.SPARK: "cpu_standard",
    Tier.CREATOR: "gpu_priority_low",
    Tier.ARTIST: "gpu_priority_medium",
    Tier.STUDIO_PRO: "gpu_priority_medium",
    Tier.ENTERPRISE: "gpu_priority_high",
}


def tier_gte(tier: str, minimum: str) -> bool:
    """Return True if `tier` has at least the rank of `minimum`."""
    try:
        t = Tier(tier)
        m = Tier(minimum)
    except ValueError:
        return False
    return TIER_RANK[t] >= TIER_RANK[m]


def queue_for_tier(tier: str) -> str:
    try:
        return TIER_QUEUE[Tier(tier)]
    except ValueError:
        return TIER_QUEUE[Tier.FREE]


def rate_limit_for_tier(tier: str) -> str:
    try:
        return TIER_RATE_LIMITS[Tier(tier)]
    except ValueError:
        return TIER_RATE_LIMITS[Tier.FREE]
